src/hex_repo.erl

```diff
@@ -148,13 +150,21 @@ get_tarball(Name, Version) ->
 %%
 %% See `get_tarball/2' for examples.
 %% @end
--spec get_tarball(string(), string(), options()) -> {ok, hex_tarball:tarball()} | {error, term()}.
+-spec get_tarball(string(), string(), options()) -> {ok, hex_tarball:tarball(), proplists:proplist()} |
+                                                    {error, term()}.
 get_tarball(Name, Version, Options) ->
     Client = proplists:get_value(client, Options),
     Repo = proplists:get_value(repo, Options),
-    case get(Client, tarball_uri(Repo, Name, Version)) of
-        {ok, {200, _Headers, Tarball}} ->
-            {ok, Tarball};
+    CacheDir = proplists:get_value(cache_dir, Options),
+
+    case get(Client, tarball_uri(Repo, Name, Version), make_headers(Options)) of
+        {ok, {200, Headers, Tarball}} ->
+            {ok, Tarball, get_headers([{<<"ETag">>, etag}], Headers)};
+        {ok, {304, _Headers, _Body}} ->
+            load_tarball_from_cache(Name, Version, CacheDir);
```

Comment:
I think we should have the same response shape whether we're hitting service or reading from cache, so we should return the etag here too.

Somewhat related: I think on cache miss, if `cache_dir` is set we should fetch & populate the cache so that it behaves as read-through cache and thus library is more useful out of the box (e.g. you don't have to manage writing to cache yourself). This makes me think however that we're putting to many responsibilities into one `get_tarball` function (and potentially all `get_*` functions because they all could use cache). I think we can keep going with this approach in this PR but we might reconsider it before first release.

Reply:
Yeah, good point about conflating responsibility too much into a single function. I'll logically split out some the cache stuff into it's own (probably embarrassingly) simple module.